src/solver/conv_hip_implicit_gemm_fwd_xdlops.cpp

```diff
@@ -188,12 +188,11 @@ namespace {
 template <typename DataType>
 void RunCKSolution(const Handle& handle,
                    const AnyInvokeParams& primitive_parameters,
-                   const ProblemDescription& problem,
-                   const PerformanceConfigHipImplicitGemmFwdXdlops& config)
+                   const CKArgs& ck_args,
+                   size_t config_idx)
 {
-    const auto args      = CKArgs{problem};
     const auto conv_ptrs = DeviceOpPtrs<DataType>::GetInstances();
-    auto& conv_ptr       = conv_ptrs.at(config.index);
+    auto& conv_ptr       = conv_ptrs.at(config_idx);
```

Comment:
Good suggestion. I didn't do this because of my lack of familiarity with CK but let me look into the lifetime of `conv_ptr` and perhaps I can capture that instead.

Reply:
@atamazov , @DrizztDoUrden : I'm posting here a combined reply to a couple of comments regarding moving `DeviceOpPtrs::GetInstances()` and moving `switch` out of the `Invoker` lambda. 

Here's why I decided not to do either:
1. The `data_type` in `ProblemDescription` is a dynamic piece of information while `ck::GetInstances()` expects a static type. Which is why we need `switch`. 
2. If I were to move the `switch` out , I'd be creating and returning a different `Invoker` lambda for each data type. Looking at the `invoker_cache.hpp` code, it seems that we don't use data type to disambiguate between Invokers of different data types for the same problem, so returning different Invokers may lead to a bug where an Invoker of one data type overwrites the cached entry of the Invoker of a different data type.

Reply:
@amberhassaan 
> we don't use data type to disambiguate between Invokers of different data types for the same problem

We use data type: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bc8891758d15d708cc3c186b9263e9a739796e3e/src/conv/problem_description.cpp#L128

Reply:
@amberhassaan We create a separate InvokerFactory for each individual Problem, this is a design choice. But kernel objects can be shared between different Invokers.

Reply:
> The data_type in ProblemDescription is a dynamic piece of information

That is not so in the context of `GetSolution` and produced invoker. The problem description is the specific problem we expect to get invoker for. Its content (including data types) is used to produce a key for `InvokerCache`. Everything which can change is stored in `*InvokeParams`. Technically speaking there are data types in tensor descriptors in invoke params, but we really do not expect every (or, I suppose, any) single kernel binary to be able to support multiple data type combinations.

At the moment we do not expect tensor descriptors to be changed as well. The main reason they are there is to allow invokers to do simple calculations while not catching inputs/results and to do fast validation.

Reply:
Thank you both for clarification. I have moved the switch statement out of the lambda and that cleans up the error logic for unsupported types as well because now `MIOPEN_THROW` is not in the invoker lambda. See 9034f3. 

Now, AFA moving the ck pointers out of the invoker lambda, such as `conv_ptr` and `invoker_ptr` , that did not work because `std::function` in the definition of `Invoker` expects the lambda to by copyable, while `conv_ptr` is only `moveable`. Please see (and try compiling) 390061 if you wish to see the compilation error.

Reply:
I had a discussion with the CK team. The are large objects and that's why they are wrapped in `unique_ptr`. What I'm going to do in a follow up PR (coming soon) is that I will cache the `unique_ptr` as a `shared_ptr` in the `invoker` lambda and then it becomes copyable and `std::function` can accept it.

Reply:
@amberhassaan 
> The are large objects and that's why they are wrapped in unique_ptr

If an object can be copied correctly, then disabling copying may be considered as a kind of over-design (unless the objects are really huge). BTW, maybe CK guys can replace unique_ptr with shared_ptr (which can be copied)?

Reply:
Yes, my plan is to convert the `unique_ptr` to `shared_ptr` in our code. `CK` doesn't need to change their return type yet.